opentelemetry-contrib/src/trace_propagator/aws_xray_propagator.rs

```diff
@@ -0,0 +1,280 @@
+/// OpenTelemetry [SpanContext][otel-spec] and [X-Ray Trace format][xray-trace-id].
+///
+/// For details on the [`x-amzn-trace-id` header][xray-header] see the AWS X-Ray Docs.
+///
+/// ## Example
+///
+/// ```
+/// extern crate opentelemetry;
+/// use opentelemetry::global;
+/// use opentelemetry_contrib::XrayTraceContextPropagator;
+///
+/// global::set_text_map_propagator(XrayTraceContextPropagator::default());
+/// ```
+///
+/// [otel-spec]: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#spancontext
+/// [xray-trace-id]: https://docs.aws.amazon.com/xray/latest/devguide/xray-api-sendingdata.html#xray-api-traceids
+/// [xray-header]: https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader
+#[derive(Clone, Debug, Default)]
+pub struct XrayTraceContextPropagator {
+    _private: (),
+}
+
+impl XrayTraceContextPropagator {
+    fn extract_span_context(&self, extractor: &dyn Extractor) -> Result<SpanContext, ()> {
+        let header_value: &str = extractor.get(AWS_XRAY_TRACE_HEADER).unwrap_or("").trim();
+
+        let parts: HashMap<&str, &str> = header_value
+            .split_terminator(';')
+            .filter_map(from_key_value_pair)
+            .collect();
+
+        let trace_id: TraceId = match parts.get(HEADER_ROOT_KEY) {
+            None => return Err(()),
+            Some(root) => {
+                let trace_id: Result<TraceId, ()> = XrayTraceId(root.to_string()).into();
+                match trace_id {
+                    Err(_) => return Err(()),
+                    Ok(trace_id) => trace_id,
+                }
+            }
+        };
+
+        let mut parent_segment_id: SpanId = SpanId::invalid();
+
+        if let Some(parent) = parts.get(HEADER_PARENT_KEY) {
+            parent_segment_id = SpanId::from_hex(parent);
+        }
+
+        let mut sampling_decision: u8 = TRACE_FLAG_DEFERRED;
+
+        if let Some(sampled_flag) = parts.get(HEADER_SAMPLED_KEY) {
+            sampling_decision = match sampled_flag.to_owned() {
+                NOT_SAMPLED => TRACE_FLAG_NOT_SAMPLED,
+                SAMPLED => TRACE_FLAG_SAMPLED,
+                REQUESTED_SAMPLE_DECISION => TRACE_FLAG_DEFERRED,
+                _ => TRACE_FLAG_DEFERRED,
+            }
+        }
+
+        Ok(SpanContext::new(
+            trace_id,
+            parent_segment_id,
+            sampling_decision,
+            true,
+        ))
+    }
+}
+
+impl TextMapFormat for XrayTraceContextPropagator {
+    fn inject_context(&self, cx: &Context, injector: &mut dyn Injector) {
+        let span_context: SpanContext = cx.span().span_context();
+        if span_context.is_valid() {
+            let xray_trace_id: XrayTraceId = span_context.trace_id().into();
+            let mut components: Vec<String> =
+                vec![format!("{}={}", HEADER_ROOT_KEY, xray_trace_id.0)];
```

Comment:
Could probably use `vec![...; 3]` to avoid possible reallocation

Reply:
So specifying `vec![...; 3]` put in three copies of the first element, refactored to use `Vec::with_capacity` instead.

Reply:
As the formatted string is always the same, you could do one allocation `injector.set(AWS_XRAY_TRACE_HEADER, format!("{}={};{}={};{}={}"))` to do one string allocation rather than 4 strings and a vec?

Reply:
Good call @jtescher, updated. Originally I had this iterating through all k/v pairs to handle arbitrary key/value pairs in the header. The [X-Ray SDK](https://github.com/aws/aws-xray-sdk-go/blob/master/header/header.go#L74) lets you pass any additional k/v pairs through, but there isn't a way to capture anything other than traceId, parent ID, and decision in the current `SpanContext` I don't think? Is that part of the thought with #195?

Reply:
Yep #195 should get support for that.